src/RamTransforms.cpp

```diff
@@ -356,41 +562,31 @@ std::unique_ptr<RamOperation> MakeIndexTransformer::rewriteIndexScan(const RamIn
     if (const auto* filter = dynamic_cast<const RamFilter*>(&iscan->getOperation())) {
         const RamRelation& rel = iscan->getRelation();
         const int identifier = iscan->getTupleId();
-        std::vector<std::unique_ptr<RamExpression>> queryPattern(rel.getArity());
+        RamPattern queryPattern;
+        for (unsigned int i = 0; i < rel.getArity(); ++i) {
+            queryPattern.first.emplace_back(nullptr);
+            queryPattern.second.emplace_back(nullptr);
+        }
+
         bool indexable = false;
         std::unique_ptr<RamCondition> condition = constructPattern(
                 queryPattern, indexable, toConjunctionList(&filter->getCondition()), identifier);
         if (indexable) {
             // Merge Index Pattern here
-            const auto prevPattern = iscan->getRangePattern();
-            auto addCondition = [&](std::unique_ptr<RamCondition> c) {
-                if (condition != nullptr) {
-                    condition = std::make_unique<RamConjunction>(std::move(condition), std::move(c));
-                } else {
-                    condition = std::move(c);
-                }
-            };
-            for (std::size_t i = 0; i < rel.getArity(); i++) {
-                if (prevPattern[i] != nullptr) {
-                    if (queryPattern[i] == nullptr) {
-                        // found a new indexable attribute
-                        queryPattern[i] = std::unique_ptr<RamExpression>(prevPattern[i]->clone());
-                    } else {
-                        // found a new constraint that is not dependent on the current scan level
-                        // and can be hoisted in a later transformation.
-                        // FIXME: `FEQ` handling; need to know if the expr is a float exp or not
-                        addCondition(std::make_unique<RamConstraint>(BinaryConstraintOp::EQ,
-                                std::unique_ptr<RamExpression>(prevPattern[i]->clone()),
-                                std::unique_ptr<RamExpression>(queryPattern[i]->clone())));
-                    }
-                }
-            }
+            RamPattern strengthenedPattern;
+            strengthenedPattern.first = clone(iscan->getRangePattern().first);
+            strengthenedPattern.second = clone(iscan->getRangePattern().second);
+
+            // strengthen the pattern with construct pattern
+            condition = constructPattern(
+                    strengthenedPattern, indexable, toConjunctionList(&filter->getCondition()), identifier);
+
```

Comment:
check for equivalence between queryPattern and strengthenedPattern; if so, return nullptr;